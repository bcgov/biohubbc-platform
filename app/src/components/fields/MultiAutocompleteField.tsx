import CheckBox from '@mui/icons-material/CheckBox';
import CheckBoxOutlineBlank from '@mui/icons-material/CheckBoxOutlineBlank';
import { Chip } from '@mui/material';
import Autocomplete, { AutocompleteInputChangeReason, createFilterOptions } from '@mui/material/Autocomplete';
import Checkbox from '@mui/material/Checkbox';
import TextField from '@mui/material/TextField';
import { useFormikContext } from 'formik';
import get from 'lodash-es/get';
import { useEffect, useState } from 'react';

export interface IMultiAutocompleteFieldOption {
  value: string | number;
  label: string;
}

export interface IMultiAutocompleteField {
  id: string;
  label: string;
  options: IMultiAutocompleteFieldOption[];
  required?: boolean;
  filterLimit?: number;
  chipVisible?: boolean;
  handleSearchResults?: (input: string) => Promise<void>;
}

export const handleSortSelectedOption = (
  selected: IMultiAutocompleteFieldOption[],
  optionsLeft: IMultiAutocompleteFieldOption[]
) => {
  const selectedOptionsValue = selected.map((item) => item.value);
  const remainingOptions = optionsLeft.filter((item) => !selectedOptionsValue.includes(item.value));

  return [...selected, ...remainingOptions];
};

const MultiAutocompleteField: React.FC<IMultiAutocompleteField> = (props) => {
  const { values, touched, errors, setFieldValue } = useFormikContext<IMultiAutocompleteFieldOption[]>();
  const [options, setOptions] = useState<IMultiAutocompleteFieldOption[]>(props.options || []); // store options if provided
  const [inputValue, setInputValue] = useState('');
  const [selectedOptions, setSelectedOptions] = useState<IMultiAutocompleteFieldOption[]>([]);

  useEffect(() => {
    setOptions(handleSortSelectedOption(selectedOptions, props.options));

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.options]);

  useEffect(() => {
    if (props.handleSearchResults) {
      props.handleSearchResults(inputValue);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [inputValue]);

  const getExistingValue = (formikValues: IMultiAutocompleteFieldOption[]): IMultiAutocompleteFieldOption[] => {
    const existingValues: IMultiAutocompleteFieldOption[] = get(formikValues, props.id);
    if (!existingValues) {
      return [];
    }

    return options.filter((option) => existingValues.includes(option));
  };

  const handleOnChange = (_event: React.ChangeEvent<any>, selectedOptions: IMultiAutocompleteFieldOption[]) => {
    setOptions(handleSortSelectedOption(selectedOptions, options));
    setSelectedOptions(selectedOptions);
    setFieldValue(
      props.id,
      selectedOptions.map((item) => item)
    );
  };

  const handleGetOptionSelected = (
    option: IMultiAutocompleteFieldOption,
    value: IMultiAutocompleteFieldOption
  ): boolean => {
    if (!option?.value || !value?.value) {
      return false;
    }

    return option.value === value.value;
  };

  const handleOnInputChange = (event: React.ChangeEvent<any>, value: string, reason: AutocompleteInputChangeReason) => {
    if (event && event.type === 'blur') {
      setInputValue('');
    } else if (reason !== 'reset') {
      setInputValue(value);
    }
  };

  const defaultChipDisplay = (option: any, renderProps: any, checkedStatus: any) => {
    return (
      <li key={option.value} {...renderProps}>
        <Checkbox
          icon={<CheckBoxOutlineBlank fontSize="small" />}
          checkedIcon={<CheckBox fontSize="small" color="primary" />}
          style={{ marginRight: 8 }}
          checked={checkedStatus}
          disabled={(options && options?.indexOf(option) !== -1) || false}
          value={option.value}
          color="default"
        />
        {option.label}
      </li>
    );
  };

  return (
    <Autocomplete
      multiple
      autoHighlight={true}
      value={getExistingValue(values)}
      id={props.id}
      options={options}
      getOptionLabel={(option) => option.label}
      isOptionEqualToValue={handleGetOptionSelected}
      filterOptions={createFilterOptions({ limit: props.filterLimit })}
      disableCloseOnSelect
      onChange={handleOnChange}
      inputValue={inputValue}
      onInputChange={handleOnInputChange}
      renderTags={(tagValue, getTagProps) => {
        if (props.chipVisible) {
          return tagValue.map((option, index) => <Chip label={option.label} {...getTagProps({ index })} />);
        }
      }}
      renderOption={(_renderProps, option, { selected }) => defaultChipDisplay(option, _renderProps, selected)}
      renderInput={(params) => (
        <TextField
          onKeyDown={(event: any) => {
            if (event.key === 'Backspace') {
              event.stopPropagation();
            }
          }}
          {...params}
          required={props.required}
          label={props.label}
          variant="outlined"
          fullWidth
          error={get(touched, props.id) && Boolean(get(errors, props.id))}
          helperText={get(touched, props.id) && get(errors, props.id)}
          placeholder={'Begin typing to filter results...'}
          InputLabelProps={{
            shrink: true
          }}
        />
      )}
    />
  );
};

export default MultiAutocompleteField;
