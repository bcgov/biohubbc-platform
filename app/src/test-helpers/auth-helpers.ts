import { SYSTEM_ROLE } from 'constants/roles';
import { IAuthState } from 'contexts/authStateContext';
import Keycloak from 'keycloak-js';

export const UnauthenticatedUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: false
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: false,
    systemRoles: [],
    hasSystemRole: () => false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    }
  }
};

export const SystemUserAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [],
    hasSystemRole: () => false,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    }
  }
};

export const SystemAdminAuthState: IAuthState = {
  keycloakWrapper: {
    keycloak: {
      authenticated: true
    } as unknown as Keycloak,
    hasLoadedAllUserInfo: true,
    systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN],
    hasSystemRole: () => true,
    getUserIdentifier: () => 'testusername',
    getIdentitySource: () => 'idir',
    username: 'testusername',
    displayName: 'testdisplayname',
    email: 'test@email.com',
    systemUserId: 1,
    refresh: () => {
      // do nothing
    },
    getLoginUrl: () => 'test/login'
  }
};

// Same effect as `Partial` but applies to all levels of a nested object
type Subset<T> = {
  [P in keyof T]?: T[P] extends Record<any, any> | undefined ? Subset<T[P]> : T[P];
};

/**
 * Build and return a mock auth state object.
 *
 * @param {{ base: IAuthState; overrides?: Subset<IAuthState> }} options
 * @return {*}  {IAuthState}
 */
export const getMockAuthState = (options: { base: IAuthState; overrides?: Subset<IAuthState> }): IAuthState => {
  const { base, overrides } = options;

  return {
    ...base,
    ...overrides,
    keycloakWrapper: {
      ...base.keycloakWrapper,
      ...overrides?.keycloakWrapper,
      Keycloak: {
        ...base.keycloakWrapper?.keycloak,
        ...overrides?.keycloakWrapper?.keycloak
      }
    }
  } as unknown as IAuthState;
};
