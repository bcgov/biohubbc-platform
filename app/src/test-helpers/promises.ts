/**
 * 
 */
export class Deferred<R = void, Q = void> {
    resolve: (value: R) => void = () => {};
    reject: (reason?: Q) => void = () => {};
    promise: Promise<R>;

    constructor() {
        this.promise = new Promise<R>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });

        return this;
    }

    recycle = () => {
        this.promise = new Promise<R>((resolve, reject) => {
            this.resolve = resolve;
            this.reject = reject;
        });

        return this;
    }
}
