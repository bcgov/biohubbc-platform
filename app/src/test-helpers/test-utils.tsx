import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@mui/material/styles'
import appTheme from 'themes/appTheme'

const AllProviders: React.FC<React.PropsWithChildren> = (props) => {
  const { children} = props
  return (
    <ThemeProvider theme={appTheme}>    
      {children}
    </ThemeProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: AllProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
