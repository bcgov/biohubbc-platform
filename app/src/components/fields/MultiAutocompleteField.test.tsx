import { render } from 'test-helpers/test-utils';
import { Formik } from 'formik';
import MultiAutocompleteField from './MultiAutocompleteField';

describe('MultiAutocompleteField', () => {
  it('renders correctly', () => {
    const { getByText } = render(
      <Formik initialValues={[]} onSubmit={async () => {}}>
        {() => (
          <MultiAutocompleteField
            id="id"
            label="label"
            options={[{ value: 'val', label: 'label' }]}
            required={true}
            filterLimit={1}
          />
        )}
      </Formik>
    );

    expect(getByText("label")).toBeVisible();
  });
});
