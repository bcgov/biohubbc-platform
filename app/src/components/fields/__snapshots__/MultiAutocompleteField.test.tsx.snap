// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiAutocompleteField matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required css-11nntsq-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="id"
        id="id-label"
      >
        label
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-uoegrp-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id="id"
          placeholder="Begin typing to filter results..."
          required=""
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1in441m"
          >
            <span>
              label *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
